File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        m, n = len(str1), len(str2)
        if m >= n:
            p, q = str1, str2
        else:
            p, q = str2, str1
        m, n = len(p), len(q)

        k = min(m, n)
        while m % k != 0 or n % k != 0:
            k -= 1
            if k == 0:
                return ""
        for i in range(k, 0, -1):
            for j in range(0, n-i+1):
                div = q[j:j+k+1]
                l = len(div)
                if p == div * (m // l) and q == div * (n // l): 
                    return div
        return ""

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.cpp
class Solution {
public:
    bool checker(string str1, string str2) {
        if (str1.length() % str2.length() != 0) return false;
        int step = str2.length();

        int cnt = str1.length() / str2.length();

        //cout << "1: " << str1 << ", " << str2 << endl;

        for (int i = 0; i < cnt; i++) {
            if (str1.substr(i*step, step).compare(str2)) {
                return false;
            }
        }

        //cout << "2: " << str1 << ", " << str2 << endl;

        return true;

    }

    string gcdOfStrings(string str1, string str2) {
        int init_cnt = str2.length();
        string div = str2;
        while (div.length() > 0) {

            if (checker(str2, div)) {
                if (checker(str1, div))
                    return div;
            }
            div.erase(init_cnt--, 1);
 
        }
        return "";

    }
};